### PR TITLE
Merge RHEL6 RPMs into `master`

### PR DESCRIPTION
**Pull request by tst-ppenev**
**Pull requests #610 and #612 need to merged before this pull request can be merged.**

* Implements build scripts that produce RPMs for Open GEE shared libraries, Server and Fusion.
* Adds documentation for building and installing RPMs.
* Fixes bugs in the Docker build.
* Adds RPM builds to the Docker build scripts.
* Updates the Docker `get-gee-build-environment.sh` to obtain a JRE  with the minimum required version.
* Removes obsolete RPM build code.
* Adds a simplistic key-value store library for exchanging values among package install scripts.
* Adds unit tests for the key-value store library.

### CODE REVIEW

**Comment by tst-ppenev** [comment]
@googlebot: I'm fine with merging my commits.

File: earth_enterprise/rpms/opengee-fusion/post-install.sh [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+    chown "$GEFUSIONUSER:$GEGROUP" "$SYSTEMRC"
+
+    if [ ! -d "$ASSET_ROOT/.config" ]; then
+        "$BASEINSTALLDIR_OPT/bin/geconfigureassetroot" --new --noprompt \
+            --assetroot "$ASSET_ROOT" --srcvol "$SOURCE_VOLUME"
+        chown -R "$GEFUSIONUSER:$GEGROUP" "$ASSET_ROOT"
+    else
+        # upgrade asset root -- if this is a master
+        if [ "$IS_SLAVE" == false ]; then
+            # TODO: Verify this logic -- this is what is defined in the
```
Comment (tst-nfarah) [inline earth_enterprise/rpms/opengee-fusion/post-install.sh]: is the 'TODO' still needed ?

File: earth_enterprise/rpms/opengee-fusion/pre-install.sh [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set +x
+
+#-----------------------------------------------------------------
+# Definitions
+GEE="Google Earth Enterprise"
```
Comment (tst-nfarah) [inline earth_enterprise/rpms/opengee-fusion/pre-install.sh]: Open Google Earth Enterprise ?

File: earth_enterprise/rpms/opengee-server/post-install.sh [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set +x
+
+#-----------------------------------------------------------------
+# Definitions
+GEE="Google Earth Enterprise"
```
Comment (tst-nfarah) [inline earth_enterprise/rpms/opengee-server/post-install.sh]: Open ... ?

File: earth_enterprise/rpms/opengee-server/post-install.sh [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+    # 5) Register publish root
+    "$BASEINSTALLDIR_OPT/bin/geconfigurepublishroot" --noprompt "--path=$PUBLISHER_ROOT"
+
+    # 6) Install the GEPlaces and SearchExample Databases
+    install_search_databases
+
+    # 7) Repeated step; Set permissions after geserver Start/Stop.
+    fix_postinstall_filepermissions
+
+    # TODO - verify
```
Comment (tst-nfarah) [inline earth_enterprise/rpms/opengee-server/post-install.sh]: Is TODO still valid ?

File: earth_enterprise/rpms/opengee-server/post-install.sh [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+    find /opt/google/share -type d -exec chmod 755 {} \;
+    find /opt/google/share -type f -exec chmod 644 {} \;
+    chmod ugo+x /opt/google/share/searchexample/searchexample
+    chmod ugo+x /opt/google/share/geplaces/geplaces
+    chmod ugo+x /opt/google/share/support/geecheck/geecheck.pl
+    chmod ugo+x /opt/google/share/support/geecheck/convert_to_kml.pl
+    chmod ugo+x /opt/google/share/support/geecheck/find_terrain_pixel.pl
+    chmod ugo+x /opt/google/share/support/geecheck/pg_check.pl
+    # Note: this is installed in install_fusion.sh, but needs setting here too.
+    #TODO
```
Comment (tst-nfarah) [inline earth_enterprise/rpms/opengee-server/post-install.sh]: remove commented code..

File: earth_enterprise/rpms/opengee-server/post-install.sh [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+    # chmod ugo+x $BASEINSTALLDIR_OPT/share/tutorials/fusion/download_tutorial.sh
+    # this should already be true...
+    chown -R root:root /opt/google/share
+
+    #TODO
```
Comment (tst-nfarah) [inline earth_enterprise/rpms/opengee-server/post-install.sh]: same as above

File: earth_enterprise/rpms/opengee-server/post-install.sh [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+    # Disable cutter (default) during installation.
+    /opt/google/bin/geserveradmin --disable_cutter
+
+    # Restrict permissions to uninstaller and installer logs
+    chmod -R go-rwx "$BASEINSTALLDIR_OPT/install"
+}
+
+reset_pgdb()
+{
+    # TODO check if correct
```
Comment (tst-nfarah) [inline earth_enterprise/rpms/opengee-server/post-install.sh]: TODO ?

File: earth_enterprise/rpms/opengee-server/post-uninstall.sh [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set +x
+
+#-----------------------------------------------------------------
+# Definitions
+GEE="Google Earth Enterprise"
```
Comment (tst-nfarah) [inline earth_enterprise/rpms/opengee-server/post-uninstall.sh]: Open GEE

File: earth_enterprise/rpms/opengee-server/pre-uninstall.sh [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set +x
+
+#-----------------------------------------------------------------
+# Definitions
+GEE="Google Earth Enterprise"
```
Comment (tst-nfarah) [inline earth_enterprise/rpms/opengee-server/pre-uninstall.sh]: Open GEE

File: earth_enterprise/docker/bin/lib/image-building.sh [hunk 10]
```diff
@@ -332,12 +334,12 @@ function tst_docker_build_dockerfile_template()
     fi
 
     DOCKER_WORKING_DIR=$(grep -Po -e "$TST_DOCKERFILE_CONTEXT_DIR_MARKER" "$DOCKERFILE_TEMPLATE_PATH")
-    if [ -n "$DOCKER_WORKING_DIR" -a "$DOCKER_CONTEXT_ARGUMENT" == "-" ]; then
+    if [ -n "$DOCKER_WORKING_DIR" -a "$DOCKER_CONTEXT_ARGUMENT" = "-" ]; then
         tst_docker_warn "Dockerfile species both no context, and context directory: $DOCKERFILE_TEMPLATE_PATH"
```
[the remark this answers is not in the document]
Comment (tst-ppenev) [reply earth_enterprise/docker/bin/lib/image-building.sh]: Thinks.  This is part of the code that's being removed.

File: earth_enterprise/docker/image-definition/Dockerfile.stage-1.clean-clone.template [hunk 11]
```diff
@@ -9,7 +9,7 @@ FROM opengee-experimental-build-env-${OS_DISTRIBUTION}
 # Build Open GEE:
 RUN cd /root/opengee/build && \
     /root/opengee/bin/pull-gee-repository.sh --branch ${CLEAN_CLONE_BRANCH_ESCAPED} ${CLEAN_CLONE_URL_ESCAPED} && \
-    /root/opengee/bin/build-gee.sh && \
+    /root/opengee/bin/build-gee.sh --packages --copy-packages && \
     /root/opengee/bin/build-portable-server.sh
 
 # We can't proceed further with installing GEE from the Dockerfile, because the
```
[the remark this answers is not in the document]
Comment (tst-ppenev) [reply earth_enterprise/docker/image-definition/Dockerfile.stage-1.clean-clone.template]: This won't be addressed on this pull request. Please, review #609.

File: earth_enterprise/docker/image-definition/Dockerfile.stage-1.current-clone.template [hunk 12]
```diff
@@ -10,8 +10,10 @@ COPY . /root/opengee/build/earthenterprise
...
     rm -Rf earth_enterprise/src/NATIVE-{REL,DBG,OPT}-x86_64 && \
-    cd .. && \
-    /root/opengee/bin/build-gee.sh && \
+    cd earth_enterprise/rpms/ && \
+    ./gradlew clean && \
+    cd ../../.. && \
+    /root/opengee/bin/build-gee.sh --packages --copy-packages && \
     /root/opengee/bin/build-portable-server.sh
 
 # We can't proceed further with installing GEE from the Dockerfile, because the
```
[the remark this answers is not in the document]
Comment (tst-ppenev) [reply earth_enterprise/docker/image-definition/Dockerfile.stage-1.current-clone.template]: Same as above.  This code is being removed.

File: earth_enterprise/rpms/opengee-fusion/post-install.sh (hunk 1 above)
Comment (tst-ppenev) [reply earth_enterprise/rpms/opengee-fusion/post-install.sh]: Yes, as far as I can tell.

This ToDo doesn't mean much to me.  I prefer a comment that says we call with `--nochown` to do this in that case, and without `--nochown` to do that in this case.

I don't know who wrote the ToDo, or the code.  You're welcome to offer a better comment.

File: earth_enterprise/rpms/opengee-fusion/pre-install.sh (hunk 2 above)
Comment (tst-ppenev) [reply earth_enterprise/rpms/opengee-fusion/pre-install.sh]: Fixed in #612.

File: earth_enterprise/rpms/opengee-server/post-install.sh (hunk 3 above)
Comment (tst-ppenev) [reply earth_enterprise/rpms/opengee-server/post-install.sh]: #612

File: earth_enterprise/rpms/opengee-server/post-install.sh (hunk 4 above)
Comment (tst-ppenev) [reply earth_enterprise/rpms/opengee-server/post-install.sh]: If you can tell me what it's supposed to be verified, I may tell you if it's still outstanding.

File: earth_enterprise/rpms/opengee-server/post-install.sh (hunk 5 above)
Comment (tst-ppenev) [reply earth_enterprise/rpms/opengee-server/post-install.sh]: Done in #612.

File: earth_enterprise/rpms/opengee-server/post-install.sh (hunk 6 above)
Comment (tst-ppenev) [reply earth_enterprise/rpms/opengee-server/post-install.sh]: This looks like something that should be fixed, but, that can be done on a clearer ticket. I'm getting rid of it.

File: earth_enterprise/rpms/opengee-server/post-install.sh (hunk 7 above)
Comment (tst-ppenev) [reply earth_enterprise/rpms/opengee-server/post-install.sh]: No sé.

File: earth_enterprise/rpms/opengee-server/post-uninstall.sh (hunk 8 above)
Comment (tst-ppenev) [reply earth_enterprise/rpms/opengee-server/post-uninstall.sh]: The unused variable has been removed in #612.

File: earth_enterprise/rpms/opengee-server/pre-uninstall.sh (hunk 9 above)
Comment (tst-ppenev) [reply earth_enterprise/rpms/opengee-server/pre-uninstall.sh]: Removed. as well.

File: earth_enterprise/BUILD_RPMS.md [hunk 13]
```diff
@@ -0,0 +1,70 @@
+# Installing via Package Managers
```
Comment (tst-nfarah) [inline earth_enterprise/BUILD_RPMS.md]: We need a section or paragraph describing how to un-install via Package Managers..

**Comment by tst-ccamp** [comment]
@tst-jcupp I think the copy-edited README changes can be made on the master branch. I don't believe it should hold up this PR

**Comment by tst-ppenev** [comment]
To all: @tst-dsugar's pull request didn't resolve the merge conflicts on this branch.  I resolved them through the Git Hub Web interface, which added commit [e216fcb](https://github.com/google/earthenterprise/pull/608/commits/e216fcbafee902e56da4102ec372d6b05a296f9c).  Please, review.

**Comment by tst-ccamp** [comment]
We need these changes in the master branch to proceed with RPM testing. All of the rest of the comments can be done as follow-ups. So, I'm going to go ahead and merge. @tst-jcupp and @tst-dsheremata any remaining requests you have for changes, please write up in a follow-up issue.

**Pull request merged**